xtask/codegen/gritql.ungram

```diff
@@ -0,0 +1,554 @@
+//   label:A     				-- suggested name for field of AST node
+
+// NOTES
+//
+// - SyntaxNode, SyntaxToken and SyntaxElement will be stripped from the codegen
+// - Bogus nodes are special nodes used to keep track of broken code; they are
+//   not part of the grammar but they will appear inside the green tree
+//
+
+
+///////////////
+// Bogus NODES
+///////////////
+// SyntaxElement is a generic data structure that is meant to track nodes and tokens
+// in cases where we care about both types
+//
+// As Bogus* node will need to yield both tokens and nodes without discrimination,
+// and their children will need to yield nodes and tokens as well.
+// For this reason, SyntaxElement = SyntaxElement
+SyntaxElement = SyntaxElement
+
+GritBogus = SyntaxElement*
+GritBogusContainer = SyntaxElement*
+GritBogusDefinition = SyntaxElement*
+GritBogusPattern = SyntaxElement*
+GritBogusLiteral = SyntaxElement*
+GritBogusNamedArg = SyntaxElement*
+GritBogusPredicate = SyntaxElement*
+
+GritRoot =
+    bom: 'UNICODE_BOM'?
+    version: GritVersion?
+    language: GritLanguageDeclaration?
+    definitions: GritDefinitionList?
+    pattern: AnyGritPattern?
+    definitions_continued: GritDefinitionList?
+    eof: 'EOF'
+
+GritSequential = 'sequential' '{' sequential: GritPatternList '}'
+
+GritFiles = 'multifile' '{' files: GritPatternList '}'
+
+AnyGritDefinition =
+    GritPatternDefinition
+    | GritPredicateDefinition
+    | GritFunctionDefinition
+    | GritBogusDefinition
+
+GritDefinitionList = AnyGritDefinition ('newline' AnyGritDefinition)* 'newline'?
+
+GritVersion = 'engine' 'biome' '(' GritDoubleLiteral ')'
+
+GritLanguageDeclaration =
+    'language'
+    name: GritLanguageName
+    flavor: GritLanguageFlavor?
+    GritBogus?
+    ';'?
+
+GritLanguageFlavor = '(' GritLanguageFlavorList ')'
+GritLanguageFlavorList = GritLanguageFlavorKind (',' GritLanguageFlavorKind)* ','?
+GritLanguageFlavorKind =
+    flavor_kind: (
+        // JavaScript flavors:
+        'typescript' | 'jsx'
+    )
+    GritBogus?
+
+// --- patterns
+
+AnyGritPattern =
+    AnyGritLiteral
+    | GritPatternNot
+    | GritPatternOr
+    | GritPatternOrElse
+    | GritPatternAny
+    | GritPatternAnd
+    | GritPatternMaybe
+    | GritPatternIfElse
+    | GritPatternContains
+    | GritPatternIncludes
+    | GritPatternAfter
+    | GritPatternBefore
+    | GritWithin
+    | GritBubble
+    | GritNodeLike
+    | GritMapAccessor
+    | GritListAccessor
+    | GritDot
+    | GritSome
+    | GritEvery
+    | GritUnderscore
+    | GritVariable
+    | GritRegexPattern
+    | GritPatternAs
+    | GritPatternLimit
+    | GritAssignmentAsPattern
+    | GritPatternAccumulate
+    | GritRewrite
+    | GritLike
+    | GritPatternWhere
+    | GritMulOperation
+    | GritDivOperation
+    | GritModOperation
+    | GritAddOperation
+    | GritSubOperation
+    | GritSequential
+    | GritFiles
+    | BracketedGritPattern
+    | GritBogusPattern
+
+BracketedGritPattern = '(' AnyGritPattern ')'
```

Comment:
All nodes should start with `Grit` unless they are known nodes that receive a special treatment in our code gen. E.g. `Any*`